### ISSUE TITLE
Synthetic data set Caffe2

### ISSUE THREAD

**Issue by markpk** [DanGraur/msc-thesis#1]
Hi Dan,

I'm trying to reproduce your caffe2 results of chapter 3 of your thesis on the DAS-5. Do you have some scripts that generate the synthetic NCHW input data that you used for your experiments?

I've some other questions as well:

1. Did you run the experiments for Caffe2 only on the CPU? That is, the files in `job-scripts/caffe2/` were not used?

2. What's the purpose of ffmpeg in your scripts? Are some of the tools it provides needed by caffe2 and TF?

3. I found some .job scripts that suggest you tried to use the CIFAR-10 data set, but in chapter 6.2 you write that you end up not using these real world data sets. Is there a particular reason for not using CIFAR-10 (since it's only < 200 MB)?

**Comment by DanGraur** [comment]
Hi @markpk,

First of all, make sure to always use the `performance_monitoring` branch and not `master`. The data should be automatically generated in the `resnet_forward_benchmark.py` script [here](https://github.com/DanGraur/msc-thesis/blob/0264dfcbb4665d943dac22b0f24e61bcb7d8d106/python/benchmark_runners/caffe2/resnet_forward_benchmark.py#L70). This script is then used by the Caffe2 [`cluster_spawner_benchmark.py`](https://github.com/DanGraur/msc-thesis/blob/performance_monitoring/python/cluster_manipulation_caffe2/cluster_spawner_benchmark/cluster_spawner_benchmark.py). The `NCHW` order is set [here](https://github.com/DanGraur/msc-thesis/blob/0264dfcbb4665d943dac22b0f24e61bcb7d8d106/python/benchmark_runners/caffe2/resnet_forward_benchmark.py#L231).

As for your other questions:

1. Yes, in the end I only ran it on CPU. I initially intended to run experiments on GPU, but that didn't work out due to time constraints. For TF I would mainly use [this](https://github.com/DanGraur/msc-thesis/blob/performance_monitoring/python/cluster_manipulation_tf/cluster_spawner_benchmark/cluster_spawner_benchmark.py) script for spawning a cluster. For Caffe2 [this](https://github.com/DanGraur/msc-thesis/blob/performance_monitoring/python/cluster_manipulation_caffe2/cluster_spawner_benchmark/cluster_spawner_benchmark.py)  script would be useful in spawning a cluster, and [this](https://github.com/DanGraur/msc-thesis/blob/0264dfcbb4665d943dac22b0f24e61bcb7d8d106/python/benchmark_runners/caffe2/resnet_forward_benchmark.py) would contain the code for the `resnet50` architecture (this can be used for both the forward and the backward experiments). I would also use the scripts in [here](https://github.com/DanGraur/msc-thesis/tree/performance_monitoring/monitoring_scripts) for monitoring hardware resources across a cluster. I would also use [this](https://github.com/DanGraur/msc-thesis/blob/performance_monitoring/python/process_killler/process_killer.py) to kill the processes started across the cluster. The `SLURM` scripts were not used in the end, because `SLURM` did not offer the required flexibility to run the experiments.  

2. I believe it is required by either Caffe2 or TF (or both). I think they use it internally for transcoding operations. From what I recall, I didn't require it directly.

3. It was indeed a small dataset, and the same applies to its counterpart `CIFAR-100`. In the end I didn't want to bias the results towards one dataset or another. `ImgNet` would've been better, since it offers quite a decent distribution over the image space, however, running one experiment on the entire dataset was not really viable from a temporal standpoint, and subsampling it would've run into the same issues that I encountered  with `CIFAR` (bias towards a particular dataset). Moreover, subsampling `ImgNet` for every new experiment would've affected result reproducibility.

Hope this helps! Let me know if you've got any other questions.

**Comment by ghost** [comment]
Hi Dan,

Thanks for your answers. I think the info in your comment should be sufficient to get started. If I have any more problems I'll let you know.

**Comment by ghost** [comment]
Hi Dan,

I managed to run the caffe2 experiments, but I get quite different timings.
I ran the experiments as follows:

```sh
#!/bin/sh
path/to/cluster_spawner_benchmark.py $# --app_path path/to/resnet_forward_benchmark.py --cpu_nodes $@ --warmup_rounds 5 --eval_rounds 10 --backward True --epoch_size 512
```

Forward + backward:

1 node: 720 ms/iter
2 nodes: 1300 ms/iter
4 nodes: 1800 ms/iter
8 nodes: 2100 ms/iter
16 nodes: 2300 ms/iter

The forward pass is around 200 ms (1 node) and 320 ms (for 2 to 16 nodes)

This is quite different from the results in Figure 3.1 and Figure 3.3 from your thesis. (I'm using Python 3.8.2 and torch 1.4.0)

However, I did get some warnings in the .out files:

`[W operator.cc:482] Shape and type inference failed for input: allreduce_0_cw for op CloneCommonWorld, skipping.` (when it's optimizing the gradient memory)

And at the end it says `Ignoring @/caffe2/caffe2/contrib/gloo:gloo_ops as it is not a valid file.` (same for the file and redis handlers)

But the benchmark (main runs) itself seems to run fine.

1. Did you get these warnings as well?

2. Did you try the MPI backend? I got a RuntimeError about an operator of type 'CopyGPUToCPU' though. So maybe it only works properly when using [`torch.distributed`](https://pytorch.org/docs/stable/distributed.html).

3. How do you control whether Ethernet or Infiniband is used? The `--network_interface` parameter of `resnet_forward_benchmark.py` doesn't seem to be used.

**Comment by DanGraur** [comment]
Hi @markpk ,

To answer your questions:

1. I don't recall getting those warnings. I used rendezvous via a file in the shared file system, and that worked well usually. I briefly tried Redis, but the solution was more complicated than the straight-forward shared file system, so I gave up on Redis. 
2. No, but that error insinuates that you're copying something from GPU memory to main memory. I never ran my experiments on GPUs, only on CPUs. Moreover, I don't recall having to interact with PyTorch (as in the end, Caffe2 and PyTorch are two different frameworks), but again, that might be because I never used GPU based training.
3. If I recall correctly, you control it by specifying InfiniBand network interface to the `--cpu_nodes` parameter, rather than using the aliases (e.g. `node053` which are bound to the Ethernet interfaces). I think you can obtain the details of the InfiniBand interface using `ip a` or `ifconfig` on the worker nodes (they are usually very regular, and only differ in the last byte of the IP address, that is they look something like this: `A.B.C.xxxx`, where `A`, `B` and `C` are constant across the nodes). You then use these addresses in the `--cpu_nodes` arg. That being said, I don't recall seeing much of a difference between Ethernet and InfinBand.

Hope this helps, and sorry if I'm a bit fuzzy on the details, but it has been about a year since I've worked on this. Let me know if there's anything else I can help out with!

**Comment by ghost** [comment]
But the timings of the results you got were obtained from output like

```
Main runs finished. Milliseconds per iter: 1774.66. Iters per second: 0.563488
```
?

I got those timings with the GLOO engine, using a rendezvous_path pointing to a folder in `/var/scratch`. I'm using the TUD nodes, btw.

**Comment by DanGraur** [comment]
The format seems familiar, but I'm not sure honestly. Do you know where in the code those lines are printed?

**Comment by ghost** [comment]
It's printed by net.cc:227 via `workspace.BenchmarkNet` (around line 566) in `resnet_forward_benchmark.py`.

**Comment by DanGraur** [comment]
Then yes, I think that should be it. I recall looking into Caffe2's `net.cc` source as well.

**Comment by ghost** [comment]
Do you happen to still have some of those .csv files mentioned in your R scripts?

**Comment by DanGraur** [comment]
I should do, yes. I'm preparing an archive with the result csvs/images/logs I managed to save from the cluster. The archive should be ready shortly.

**Comment by DanGraur** [comment]
The archive can be downloaded from [here](https://polybox.ethz.ch/index.php/s/9lUj7CcexGXtmRH/download). The folder structure is quite messy, so I apologize for that, but hopefully you can interpolate from the R scripts which .csvs I used.

**Comment by ghost** [comment]
Another question: the timing printed by the benchmark script is per machine right? Shouldn't it decrease when more nodes are used?

**Comment by ghost** [comment]
It seems the global size is increased as a function of the number of nodes, so it's correct to say the timings printed to the .out files should stay the same when you assume there is no extra communication overhead?

**Comment by DanGraur** [comment]
By global size do you mean global batch? If so, yes. And also yes, I think that should be the case.

**Comment by DanGraur** [comment]
I will close this issue for now. Feel free to open it again when you have other questions.
